src/slurm_cloud_bursting/suspend.py

```diff
@@ -0,0 +1,174 @@
+    max_retry = int(config.get("slurm_cb_config", "max_retry", fallback=DEFAULT_MAX_RETRY))
+    if max_retry > 0:
+        boto3_config["retries"] = {
+            "max_attempts": max_retry,
+            "mode": "standard",
+        }
+    max_batch_size = int(config.get("slurm_cb_config", "max_batch_size", fallback=DEFAULT_MAX_INSTANCES_BATCH_SIZE))
+
+    proxy = config.get("slurm_cb_config", "proxy", fallback="NONE")
+    if proxy != "NONE":
+        boto3_config["proxies"] = {"https": proxy}
+
+    logging.info(
+        (
+            "Configured parameters: region=%s cluster_name=%s boto3_config=%s max_batch_size=%s",
+            region,
+            cluster_name,
+            boto3_config,
+            max_retry,
+            max_batch_size,
+        )
+    )
+    return SlurmSuspendConfig(region, cluster_name, Config(**boto3_config), max_batch_size)
+
+
+def delete_instances(instance_ids_to_nodename, region, boto3_config, batch_size):
+    """Terminate corresponding EC2 instances."""
+    ec2_client = boto3.client("ec2", region_name=region, config=boto3_config)
+    logging.info("Terminating instances %s", list(instance_ids_to_nodename.keys()))
+    for instances in grouper(instance_ids_to_nodename.keys(), batch_size):
+        try:
+            ec2_client.terminate_instances(InstanceIds=instances,)
+        except ClientError as e:
+            logging.error("Failed when terminating instances %s with error %d", instances, e)
+            _set_nodes_to_down([instance_ids_to_nodename[instance_id] for instance_id in instances])
+
+
+def _get_instance_ids_to_nodename(slurm_nodes, region, boto3_config):
+    """Retrieve dict that maps from instance ids to slurm nodenames."""
+    node_ip_to_name = {node.nodeaddr: node.name for node in slurm_nodes}
+    ec2_client = boto3.client("ec2", region_name=region, config=boto3_config)
+    paginator = ec2_client.get_paginator("describe_instances")
+    response_iterator = paginator.paginate(
+        Filters=[{"Name": "private-ip-address", "Values": list(node_ip_to_name.keys())}],
+    )
+    filtered_iterator = response_iterator.search("Reservations[].Instances[]")
+    return {
+        instance_info["InstanceId"]: node_ip_to_name[instance_info["PrivateIpAddress"]]
+        for instance_info in filtered_iterator
+    }
+
+
+def _set_nodes_to_idle(slurm_nodes):
+    """Set POWER_DOWN nodes back to IDLE by resuming the node."""
+    logging.info("Resuming the following nodes back to IDLE: %s", slurm_nodes)
+    for node in slurm_nodes:
+        resume_nodes_command = (
+            f"{SCONTROL} update nodename={node.name} nodehostname={node.name} nodeaddr={node.name} state=resume"
+        )
+        # Do not raise on error.
+        # Failure for resume command will fail if node already in IDLE, ignore failure.
+        run_command(resume_nodes_command, raise_on_error=False)
+
+
+def _set_nodes_to_down(slurm_nodes):
+    """
+    Fall back mechanism to handle failure when removing instances.
```

Comment:
when a node is in powering down state it's not used by the scheduler until it is resumed. Rather than setting it as down I would simply resume the node.